lib/aws-manager.js

```diff
@@ -81,142 +82,162 @@ class AwsManager {
-        ],
-      });
-      debug('fetched dead instance state');
+    this.__previousApiState = this.__apiState;
+    let apiState = {
+      instances: [],
+      requests: [],
+    };
+    let deadState = {
+      instances: [],
+      requests: [],
+    };
 
-      let rawDeadSR = await this.ec2.describeSpotInstanceRequests({
-        Filters: [
-          {
-            Name: 'launch.key-name',
-            Values: [this.keyPrefix + '*'],
-          }, {
-            Name: 'state',
-            Values: ['cancelled', 'failed', 'closed', 'active'],
-          },
-        ],
-      });
-      debug('fetched dead spot request state');
+    let stalledKills = [];
+    let availableAZ = {};
+    let allPricingHistory = {};
+
+    // It would be better to do it by classification than by region for
+    // better concurrency, but this is easier and not too slow considering
+    // the 75s iteration frequency
+    await Promise.all(_.map(this.ec2, async (ec2, region) => {
+      let response = await Promise.all([
+        // Living instances
+        ec2.describeInstances({
+          Filters: [
+            {
+              Name: 'key-name',
+              Values: [this.keyPrefix + '*'],
+            },
+            {
+              Name: 'instance-state-name',
+              Values: ['running', 'pending'],
+            },
+          ],
+        }).promise(),
+        // Living spot requests
+        ec2.describeSpotInstanceRequests({
+          Filters: [
+            {
+              Name: 'launch.key-name',
+              Values: [this.keyPrefix + '*'],
+            }, {
+              Name: 'state',
+              Values: ['open'],
+            },
+          ],
+        }).promise(),
+        // Dead instances
+        ec2.describeInstances({
+          Filters: [
+            {
+              Name: 'key-name',
+              Values: [this.keyPrefix + '*'],
+            },
+            {
+              Name: 'instance-state-name',
+              Values: ['shutting-down', 'terminated', 'stopping'],
+            },
+          ],
+        }).promise(),
+        // Dead spot requests
+        ec2.describeSpotInstanceRequests({
+          Filters: [
+            {
+              Name: 'launch.key-name',
+              Values: [this.keyPrefix + '*'],
+            }, {
+              Name: 'state',
+              Values: ['cancelled', 'failed', 'closed', 'active'],
+            },
+          ],
+        }).promise(),
+        // Available availability zones
+        ec2.describeAvailabilityZones({
+          Filters: [
+            {
+              Name: 'state',
+              Values: ['available'],
+            },
+          ],
+        }).promise(),
+        // Raw pricing data
+        ec2.describeSpotPriceHistory({
+          StartTime: pricingStartDate,
+          Filters: [
+            {
+              Name: 'product-description',
+              Values: ['Linux/UNIX'],
+            },
+          ],
+        }).promise(),
+      ]);
```

Comment:
I still propose refactoring this into methods and combine the result with `_.flatten` and `_.assign` like:
```js
[
  apiState.instances,
  apiState.requests,
  ...
] = await Promise.all([
  Promise.all(_.map(this.ec2, region => this._fetchLivingInstances)).then(_.flatten),
  Promise.all(_.map(this.ec2, region => this._fetchLivingRequests)).then(_.flatten),
  ...
]);
```
See comment: https://github.com/taskcluster/aws-provisioner/pull/55#discussion_r53674042

----
It's probably not blocking, but seriously consider it.